cmd/containerboot/main.go

```diff
@@ -0,0 +1,423 @@
+//
+//  - TS_AUTH_KEY: the authkey to use for login.
+//  - TS_ROUTES: subnet routes to advertise.
+//  - TS_DEST_IP: proxy all incoming Tailscale traffic to the given
+//                destination.
+//  - TS_TAILSCALED_EXTRA_ARGS: extra arguments to 'tailscaled'.
+//  - TS_EXTRA_ARGS: extra arguments to 'tailscale up'.
+//  - TS_USERSPACE: run with userspace networking (the default)
+//                  instead of kernel networking.
+//  - TS_STATE_DIR: the directory in which to store tailscaled
+//                  state. The data should persist across container
+//                  restarts.
+//  - TS_ACCEPT_DNS: whether to use the tailnet's DNS configuration.
+//  - TS_KUBE_SECRET: the name of the Kubernetes secret in which to
+//                    store tailscaled state.
+//  - TS_SOCKS5_SERVER: the address on which to listen for SOCKS5
+//                      proxying into the tailnet.
+//  - TS_OUTBOUND_HTTP_PROXY_LISTEN: the address on which to listen
+//                                   for HTTP proxying into the tailnet.
+//  - TS_SOCKET: the path where the tailscaled local API socket should
+//               be created.
+//  - TS_AUTH_ONCE: if true, only attempt to log in if not already
+//                  logged in. If false (the default, for backwards
+//                  compatibility), forcibly log in every time the
+//                  container starts.
+//
+// When running on Kubernetes, TS_KUBE_SECRET takes precedence over
+// TS_STATE_DIR. Additionally, if TS_AUTH_KEY is not provided and the
+// TS_KUBE_SECRET contains an "authkey" field, that key is used.
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"net/netip"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+	"tailscale.com/client/tailscale"
+	"tailscale.com/ipn/ipnstate"
+)
+
+func main() {
+	log.SetPrefix("boot: ")
+	tailscale.I_Acknowledge_This_API_Is_Unstable = true
+
+	cfg := &settings{
+		AuthKey:         defaultEnv("TS_AUTH_KEY", ""),
+		Routes:          defaultEnv("TS_ROUTES", ""),
+		ProxyTo:         defaultEnv("TS_DEST_IP", ""),
+		DaemonExtraArgs: defaultEnv("TS_TAILSCALED_EXTRA_ARGS", ""),
+		ExtraArgs:       defaultEnv("TS_EXTRA_ARGS", ""),
+		InKubernetes:    os.Getenv("KUBERNETES_SERVICE_HOST") != "",
+		UserspaceMode:   defaultBool("TS_USERSPACE", true),
+		StateDir:        defaultEnv("TS_STATE_DIR", ""),
+		AcceptDNS:       defaultBool("TS_ACCEPT_DNS", false),
+		KubeSecret:      defaultEnv("TS_KUBE_SECRET", "tailscale"),
+		SOCKSProxyAddr:  defaultEnv("TS_SOCKS5_SERVER", ""),
+		HTTPProxyAddr:   defaultEnv("TS_OUTBOUND_HTTP_PROXY_LISTEN", ""),
+		Socket:          defaultEnv("TS_SOCKET", "/tmp/tailscaled.sock"),
+		AuthOnce:        defaultBool("TS_AUTH_ONCE", false),
+	}
+
+	if cfg.ProxyTo != "" && cfg.UserspaceMode {
+		log.Fatal("TS_DEST_IP is not supported with TS_USERSPACE")
+	}
+	if os.Getenv("TS_KUBE_SECRET") != "" && !cfg.InKubernetes {
```

Comment:
nit: add a comment saying that we want to check the env var directly here in case someone accidentally refactors this in the future.

Reply:
I actually removed this test entirely, because technically it's breaking compat with run.sh: run.sh was completely happy to let you set TS_KUBE_SECRET, and just ignored it if not running in k8s. I could imagine that people have things set up to unconditionally pass TS_KUBE_SECRET, but then use that config both in docker for local dev and in k8s in prod. So, even though it might allow some buggy configs, I think we need to keep allowing it :(

Reply:
oh sure, bug for bug compat sgtm